File: packages/core/src/state.ts
// Game state: players, zones (piles), globals, RNG, and visibility-aware
// observations.

import { Card, Pile, Player, CSRecord } from "./values.ts";
import type { CSValue, ZoneDef, ZoneHandle } from "./values.ts";
import { RNG } from "./rng.ts";

// deep-clone a game-global value, remapping player references to the clone's
// players. Cards are shared (identity by id). Zone handles / callables aren't
// stored in game globals by any current game, so they're shared as-is.
function cloneValue(v: CSValue, pmap: Map<number, Player>): CSValue {
  if (v === null || typeof v !== "object") return v;
  if (v instanceof Card) return v;
  if (v instanceof Player) return pmap.get(v.id) ?? v;
  if (Array.isArray(v)) return v.map((x) => cloneValue(x, pmap));
  if (v instanceof CSRecord) {
    const r = new CSRecord();
    for (const [k, val] of v.map) r.set(k, cloneValue(val, pmap));
    return r;
  }
  return v;
}

export class GameState {
  players: Player[];
  rng: RNG;

  zoneDefs = new Map<string, ZoneDef>();
  sharedPiles = new Map<string, Pile>();
  perPlayerPiles = new Map<string, Pile[]>();

  current: Player;
  ended = false;
  declaredWinners: Player[] | null = null;
  turnCount = 0;
  private nextCardId = 1;

  // game-global variables (var declarations at game scope, plus engine vars)
  globals = new Map<string, CSValue>();

  // optional host sink for announce() narration (UI move log, CLI, etc.)
  onAnnounce: ((msg: string) => void) | null = null;

  constructor(numPlayers: number, seed: number, names?: string[]) {
    this.players = [];
    for (let i = 0; i < numPlayers; i++) {
      this.players.push(new Player(i, names?.[i] ?? `P${i + 1}`));
    }
    this.rng = new RNG(seed);
    this.current = this.players[0];
  }

  // ---- zones ----
  defineZone(def: ZoneDef): void {
    this.zoneDefs.set(def.name, def);
    if (def.perPlayer) {
      this.perPlayerPiles.set(def.name, this.players.map((p) => new Pile(def, p)));
    } else {
      this.sharedPiles.set(def.name, new Pile(def, null));
    }
  }

  zoneHandle(name: string): ZoneHandle | null {
    const def = this.zoneDefs.get(name);
    if (!def) return null;
    if (def.perPlayer) {
      return { zone: "family", def, piles: this.perPlayerPiles.get(name) as Pile[] };
    }
    return { zone: "pile", pile: this.sharedPiles.get(name) as Pile };
  }

  pileOf(name: string, player?: Player): Pile | null {
    const def = this.zoneDefs.get(name);
    if (!def) return null;
    if (def.perPlayer) {
      if (!player) return null;
      return (this.perPlayerPiles.get(name) as Pile[])[player.id];
    }
    return this.sharedPiles.get(name) as Pile;
  }

  // ---- cloning (for snapshot / search / determinization) ----
  // A deep, independent copy: players and piles are fresh objects, Card objects
  // are shared (immutable identity — `id` equality must hold across the clone),
  // and game globals are deep-cloned with player references remapped. zoneDefs
  // are immutable and shared. The clone is silent (onAnnounce = null).
  clone(): GameState {
    const c = new GameState(this.players.length, 0);
    c.players = this.players.map((p) => {
      const np = new Player(p.id, p.name);
      np.eliminated = p.eliminated;
      return np;
    });
    const pmap = new Map(c.players.map((p) => [p.id, p]));

    c.zoneDefs = this.zoneDefs; // immutable
    c.sharedPiles = new Map();
    for (const [name, pile] of this.sharedPiles) {
      const np = new Pile(pile.def, null);
      np.cards = pile.cards.slice();
      c.sharedPiles.set(name, np);
    }
    c.perPlayerPiles = new Map();
    for (const [name, piles] of this.perPlayerPiles) {
      c.perPlayerPiles.set(
        name,
        piles.map((pile) => {
          const np = new Pile(pile.def, pile.owner ? (pmap.get(pile.owner.id) as Player) : null);
          np.cards = pile.cards.slice();
          return np;
        }),
      );
    }

    c.current = pmap.get(this.current.id) as Player;
    c.ended = this.ended;
    c.declaredWinners = this.declaredWinners
      ? this.declaredWinners.map((p) => pmap.get(p.id) as Player)
      : null;
    c.turnCount = this.turnCount;
    c.nextCardId = this.nextCardId;
    c.rng = this.rng.clone();
    c.globals = new Map();
    for (const [k, v] of this.globals) c.globals.set(k, cloneValue(v, pmap));
    c.onAnnounce = null;
    return c;
  }

  private canSeePile(pile: Pile, viewerId: number): boolean {
    switch (pile.def.visibility) {
      case "up":
        return true;
      case "down":
        return false;
      case "owner":
        return pile.owner?.id === viewerId;
    }
  }

  // A determinization for `viewer`: a clone in which every card the viewer can
  // NOT see is randomly re-dealt among the hidden slots (counts and visible
  // cards preserved). This is the fair input to information-set search — the
  // searcher reasons over plausible worlds consistent with its observation, and
  // never sees the true hidden layout or the game seed. Sampling uses `rng`
  // (a fresh, search-local RNG — not the game's).
  determinize(viewer: Player, rng: RNG): GameState {
    const c = this.clone();
    const hiddenPiles: Pile[] = [];
    const pool: Card[] = [];
    const collect = (pile: Pile): void => {
      if (!c.canSeePile(pile, viewer.id)) {
        hiddenPiles.push(pile);
        for (const card of pile.cards) pool.push(card);
      }
    };
    for (const pile of c.sharedPiles.values()) collect(pile);
    for (const piles of c.perPlayerPiles.values()) for (const pile of piles) collect(pile);

    rng.shuffle(pool);
    let i = 0;
    for (const pile of hiddenPiles) {
      for (let k = 0; k < pile.cards.length; k++) pile.cards[k] = pool[i++];
    }
    return c;
  }

  // ---- deck ----
  buildStandard52(into: Pile): void {
    for (let suit = 0; suit < 4; suit++) {
      for (let rank = 1; rank <= 13; rank++) {
        into.cards.push(new Card(rank, suit, rank, this.nextCardId++));
      }
    }
  }

  // ---- players / turns ----
  activePlayers(): Player[] {
    return this.players.filter((p) => !p.eliminated);
  }
  nextActiveAfter(p: Player): Player {
    const n = this.players.length;
    for (let k = 1; k <= n; k++) {
      const cand = this.players[(p.id + k) % n];
      if (!cand.eliminated) return cand;
    }
    return p;
  }

  // ---- visibility-aware observation (for networking / ML) ----
  // Returns a plain JSON-able snapshot from `viewer`'s perspective. Cards in
  // zones the viewer may not see are replaced by `null` placeholders, so the
  // count is known but identities are hidden.
  observe(viewer: Player): Observation {
    const zones: Record<string, ZoneView | ZoneView[]> = {};
    for (const [name, def] of this.zoneDefs) {
      if (def.perPlayer) {
        zones[name] = (this.perPlayerPiles.get(name) as Pile[]).map((pile) =>
          this.viewPile(pile, viewer),
        );
      } else {
        zones[name] = this.viewPile(this.sharedPiles.get(name) as Pile, viewer);
      }
    }
    return {
      viewer: viewer.id,
      current: this.current.id,
      players: this.players.map((p) => ({ id: p.id, name: p.name, out: p.eliminated })),
      zones,
      turn: this.turnCount,
    };
  }

  private canSee(pile: Pile, viewer: Player): boolean {
    switch (pile.def.visibility) {
      case "up":
        return true;
      case "down":
        return false;
      case "owner":
        return pile.owner?.id === viewer.id;
    }
  }

  private viewPile(pile: Pile, viewer: Player): ZoneView {
    const visible = this.canSee(pile, viewer);
    return {
      size: pile.cards.length,
      owner: pile.owner ? pile.owner.id : null,
      layout: pile.def.layout,
      cards: pile.cards.map((c) =>
        visible ? { rank: c.rank, suit: c.suit, value: c.value, id: c.id } : null,
      ),
    };
  }
}

export interface ZoneView {
  size: number;
  owner: number | null;
  layout: "pile" | "hand";
  cards: ({ rank: number; suit: number; value: number; id: number } | null)[];
}

export interface Observation {
  viewer: number;
  current: number;
  players: { id: number; name: string; out: boolean }[];
  zones: Record<string, ZoneView | ZoneView[]>;
  turn: number;
}

File: packages/pyengine/cardsharp/state.py
"""Game state: players, zones (piles), globals, RNG, visibility-aware
observations. Port of packages/core/src/state.ts."""

from .values import Card, Pile, Player, ZoneDef, ZoneHandle, CSRecord
from .rng import RNG


def _clone_value(v, pmap):
    if v is None or not isinstance(v, (Card, Player, list, CSRecord)):
        return v
    if isinstance(v, Card):
        return v
    if isinstance(v, Player):
        return pmap.get(v.id, v)
    if isinstance(v, list):
        return [_clone_value(x, pmap) for x in v]
    if isinstance(v, CSRecord):
        r = CSRecord()
        for k, val in v.map.items():
            r.set(k, _clone_value(val, pmap))
        return r
    return v


class GameState:
    def __init__(self, num_players, seed, names=None):
        self.players = [Player(i, (names[i] if names and i < len(names) else f"P{i + 1}"))
                        for i in range(num_players)]
        self.rng = RNG(seed)
        self.zone_defs = {}        # name -> ZoneDef
        self.shared_piles = {}     # name -> Pile
        self.per_player_piles = {}  # name -> [Pile]
        self.current = self.players[0]
        self.ended = False
        self.declared_winners = None
        self.turn_count = 0
        self._next_card_id = 1
        self.globals = {}
        self.on_announce = None

    # ---- zones ----
    def define_zone(self, zdef):
        self.zone_defs[zdef.name] = zdef
        if zdef.perPlayer:
            self.per_player_piles[zdef.name] = [Pile(zdef, p) for p in self.players]
        else:
            self.shared_piles[zdef.name] = Pile(zdef, None)

    def zone_handle(self, name):
        zdef = self.zone_defs.get(name)
        if not zdef:
            return None
        if zdef.perPlayer:
            return ZoneHandle("family", zdef=zdef, piles=self.per_player_piles[name])
        return ZoneHandle("pile", pile=self.shared_piles[name])

    def pile_of(self, name, player=None):
        zdef = self.zone_defs.get(name)
        if not zdef:
            return None
        if zdef.perPlayer:
            if player is None:
                return None
            return self.per_player_piles[name][player.id]
        return self.shared_piles[name]

    # ---- cloning ----
    def clone(self):
        c = GameState(len(self.players), 0)
        c.players = []
        for p in self.players:
            np = Player(p.id, p.name)
            np.eliminated = p.eliminated
            c.players.append(np)
        pmap = {p.id: p for p in c.players}

        c.zone_defs = self.zone_defs  # immutable
        c.shared_piles = {}
        for name, pile in self.shared_piles.items():
            np = Pile(pile.zdef, None)
            np.cards = list(pile.cards)
            c.shared_piles[name] = np
        c.per_player_piles = {}
        for name, piles in self.per_player_piles.items():
            newpiles = []
            for pile in piles:
                owner = pmap.get(pile.owner.id) if pile.owner else None
                np = Pile(pile.zdef, owner)
                np.cards = list(pile.cards)
                newpiles.append(np)
            c.per_player_piles[name] = newpiles

        c.current = pmap[self.current.id]
        c.ended = self.ended
        c.declared_winners = ([pmap[p.id] for p in self.declared_winners]
                              if self.declared_winners is not None else None)
        c.turn_count = self.turn_count
        c._next_card_id = self._next_card_id
        c.rng = self.rng.clone()
        c.globals = {k: _clone_value(v, pmap) for k, v in self.globals.items()}
        c.on_announce = None
        return c

    def _can_see_pile(self, pile, viewer_id):
        vis = pile.zdef.visibility
        if vis == "up":
            return True
        if vis == "down":
            return False
        return pile.owner is not None and pile.owner.id == viewer_id

    def determinize(self, viewer, rng):
        c = self.clone()
        hidden = []
        pool = []
        for pile in c.shared_piles.values():
            if not c._can_see_pile(pile, viewer.id):
                hidden.append(pile)
                pool.extend(pile.cards)
        for piles in c.per_player_piles.values():
            for pile in piles:
                if not c._can_see_pile(pile, viewer.id):
                    hidden.append(pile)
                    pool.extend(pile.cards)
        rng.shuffle(pool)
        i = 0
        for pile in hidden:
            for k in range(len(pile.cards)):
                pile.cards[k] = pool[i]
                i += 1
        return c

    def determinize_in_place(self, viewer, rng):
        """Reshuffle the cards the viewer cannot see, mutating the existing pile
        arrays. Used to determinize a cloned Machine whose frames already
        reference these pile objects."""
        hidden = []
        pool = []
        for pile in self.shared_piles.values():
            if not self._can_see_pile(pile, viewer.id):
                hidden.append(pile)
                pool.extend(pile.cards)
        for piles in self.per_player_piles.values():
            for pile in piles:
                if not self._can_see_pile(pile, viewer.id):
                    hidden.append(pile)
                    pool.extend(pile.cards)
        rng.shuffle(pool)
        i = 0
        for pile in hidden:
            for k in range(len(pile.cards)):
                pile.cards[k] = pool[i]
                i += 1

    # ---- deck ----
    def build_standard52(self, into):
        for suit in range(4):
            for rank in range(1, 14):
                into.cards.append(Card(rank, suit, rank, self._next_card_id))
                self._next_card_id += 1

    # ---- players / turns ----
    def active_players(self):
        return [p for p in self.players if not p.eliminated]

    def next_active_after(self, p):
        n = len(self.players)
        for k in range(1, n + 1):
            cand = self.players[(p.id + k) % n]
            if not cand.eliminated:
                return cand
        return p

    # ---- observation ----
    def observe(self, viewer):
        zones = {}
        for name, zdef in self.zone_defs.items():
            if zdef.perPlayer:
                zones[name] = [self._view_pile(pile, viewer)
                               for pile in self.per_player_piles[name]]
            else:
                zones[name] = self._view_pile(self.shared_piles[name], viewer)
        return {
            "viewer": viewer.id,
            "current": self.current.id,
            "players": [{"id": p.id, "name": p.name, "out": p.eliminated}
                        for p in self.players],
            "zones": zones,
            "turn": self.turn_count,
        }

    def _view_pile(self, pile, viewer):
        visible = self._can_see_pile(pile, viewer.id)
        return {
            "size": len(pile.cards),
            "owner": pile.owner.id if pile.owner else None,
            "layout": pile.zdef.layout,
            "cards": [({"rank": c.rank, "suit": c.suit, "value": c.value, "id": c.id}
                       if visible else None) for c in pile.cards],
        }
